README.rst
|pypi| |travis-ci| |codecov| |docs|

aioethereum
===========

Ethereum RPC client library for the `PEP 3156`_ Python event loop.

.. _PEP 3156: http://legacy.python.org/dev/peps/pep-3156/

Features
--------

================================  ==============================
ujson_ support                      Yes
uvloop_ support                     Yes
High-level APIs                     Yes
HTTP support                        Yes
Unix domain socket (IPC) support    Yes
SSL/TLS support                     Yes
Tested CPython versions             `3.4, 3.5, 3.6 <travis_>`_
Tested Geth versions                `1.7.0 <travis_>`_
Implemented RPC apis                `admin, db, debug, eth, miner, net, personal, shh, txpool, web3`
================================  ==============================

Documentation
-------------

http://aioethereum.readthedocs.io/

Usage examples
--------------

Simple high-level interface (through HTTP):

.. code:: python

    import asyncio
    import aioethereum

    loop = asyncio.get_event_loop()

    async def go():
        client = await aioethereum.create_ethereum_client(
            'http://localhost:8545', loop=loop)
        val = await client.web3_clientVersion()
        print(val)

    loop.run_until_complete(go())
    # will print like 'Geth/v1.7.0-stable-6c6c7b2a/darwin-amd64/go1.9'


or via IPC

.. code:: python

    import asyncio
    import aioethereum

    loop = asyncio.get_event_loop()

    async def go():
        client = await aioethereum.create_ethereum_client(
            'ipc://<path_to_unix_socket>', loop=loop)
        val = await client.web3_clientVersion()
        print(val)

    loop.run_until_complete(go())
    # will print like 'Geth/v1.7.0-stable-6c6c7b2a/darwin-amd64/go1.9'


Requirements
------------

* Python_ 3.3+
* asyncio_ or Python_ 3.4+
* ujson_
* aiohttp_

.. note::

    ujson is preferred requirement.
    Pure C JSON encoder and decoder is implemented as well and can be used
    automatically when installed.


License
-------

The aioethereum is offered under MIT license.

.. _Python: https://www.python.org
.. _asyncio: https://pypi.python.org/pypi/asyncio
.. _aiohttp: https://pypi.python.org/pypi/aiohttp
.. _ujson: https://pypi.python.org/pypi/ujson
.. _uvloop: https://pypi.python.org/pypi/uvloop
.. _travis: https://travis-ci.org/DeV1doR/aioethereum


.. |pypi| image:: https://badge.fury.io/py/aioethereum.svg
    :target: https://badge.fury.io/py/aioethereum
    :alt: Latest version released on PyPi

.. |travis-ci| image:: https://travis-ci.org/DeV1doR/aioethereum.svg?branch=master
    :target: https://travis-ci.org/DeV1doR/aioethereum
    :alt: Travis CI status

.. |docs| image:: https://readthedocs.org/projects/aioethereum/badge/?version=stable
    :target: https://aioethereum.readthedocs.io/
    :alt: Documentation status

.. |codecov| image:: https://codecov.io/gh/DeV1doR/aioethereum/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/DeV1doR/aioethereum
    :alt: Test coverage

.. |license| image:: https://img.shields.io/pypi/l/aioethereum.svg?style=flat&label=license
    :target: https://github.com/DeV1doR/aioethereum/blob/master/LICENSE.md
    :alt: MIT License


CHANGES.rst
0.2.2 (2018-04-10)
^^^^^^^^^^^^^^^^^^

* Fix bug, related to https://www.python.org/dev/peps/pep-0492/#new-syntax;
* Fix port detection for client when only domain;


0.2.1 (2017-10-08)
^^^^^^^^^^^^^^^^^^

* Add admin and debug management apis;
* Add new tests;
* Add uvloop support (python 3.5+ required);


0.2.0 (2017-10-05)
^^^^^^^^^^^^^^^^^^

* Add more docstrings to the code;
* Add tests for all rpc methods;
* Add admin and debug;
* Fix error for unixsocket retring;
* Fix unixsocket invalid loop for Python 3.4;
* BaseAsyncIOClient._rpc marked as deprecated and will be removed in 0.3.0;


0.1.1 (2017-10-01)
^^^^^^^^^^^^^^^^^^

* Add sphinx docs;


0.1.0 (2017-09-30)
^^^^^^^^^^^^^^^^^^

* Initial release;
* Ethereum client implemented;
* WIP on RPC management.

aioethereum/client.py
import logging
import asyncio
import warnings
import abc
import sys
from urllib.parse import urlparse

try:
    from abc import ABC
except ImportError:
    class ABC(metaclass=abc.ABCMeta):
        pass

import aiohttp
import async_timeout

try:
    import uvloop
    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
except ImportError:
    pass

try:
    import ujson as json  # noqa
except ImportError:
    import json

from .errors import BadResponseError, BadStatusError, BadJsonError
from .management import RpcMixin


logger = logging.getLogger('asyncio_client')


_reconnect_times = 3


class BaseAsyncIOClient(ABC):
    """Abstract class for creating client.
    """

    @abc.abstractmethod
    def rpc_call(self, method, params=None, id_=None):
        """Implements RPC 2.0 call to node server.

        :param method: RPC node method
        :type method: str

        :param params: Input params for method
        :type params: list

        :param id_: ID of current request
        :type id_: str or int

        :return: `Response from RPC`
        """

    @asyncio.coroutine
    def _call(self, method, params=None, id_=None):
        warnings.warn('This method is deprecated and will be removed '
                      'in v0.3.0', DeprecationWarning)
        return (yield from self.rpc_call(method, params, id_))


class AsyncIOHTTPClient(BaseAsyncIOClient, RpcMixin):
    """Creates AsyncIOHTTPClient client to communicate via HTTP(s).

    :param host: Host on ethereum node
    :type host: str

    :param port: Port on ethereum node
    :type port: int

    :param tls: Use SSL connection
    :type tls: bool

    :param timeout: Total time of timeout call
    :type timeout: int

    :param loop: An optional *event loop* instance
                 (uses :func:`asyncio.get_event_loop` if not specified).
    :type loop: :ref:`EventLoop<asyncio-event-loop>`

    :return: :class:`AsyncIOHTTPClient` instance.
    """

    def __init__(self, host='127.0.0.1', port=8545, tls=False,
                 timeout=60, *, loop=None):
        self._host = host
        self._port = port
        self._tls = tls
        self._timeout = timeout
        self._id = 1
        self._loop = loop or asyncio.get_event_loop()

    @property
    def _endpoint(self):
        scheme = 'http'
        if self._tls:
            scheme += 's'
        return '{0}://{1}:{2}'.format(scheme, self._host, self._port)

    @asyncio.coroutine
    def _post(self, data):

        # Newer versions of aiohttp raise TypeError if ClientSession is used
        # in a "non-async with" statement. This method is a workaround.
        # See https://www.python.org/dev/peps/pep-0492/#new-syntax

        contextmanager = aiohttp.ClientSession(loop=self._loop)
        try:
            session = yield from contextmanager.__aenter__()
        except AttributeError:
            session = contextmanager.__enter__()
        try:
            with async_timeout.timeout(self._timeout, loop=self._loop):
                # do not return just yet, the "else" branch needs to run too
                r = yield from session.post(
                    url=self._endpoint,
                    data=json.dumps(data),
                    headers={'Content-Type': 'application/json'}
                )
        except:  # noqa
            try:
                exit = yield from contextmanager.__aexit__(*sys.exc_info())
            except AttributeError:
                exit = contextmanager.__exit__(*sys.exc_info())

            if not exit:
                raise
        else:
            try:
                yield from contextmanager.__aexit__(None, None, None)
            except AttributeError:
                contextmanager.__exit__(None, None, None)
        return r

    @asyncio.coroutine
    def rpc_call(self, method, params=None, id_=None):
        params = params or []
        data = {
            'jsonrpc': '2.0',
            'method': method,
            'params': params,
            'id': id_ or self._id,
        }
        try:
            r = yield from self._post(data)
        except aiohttp.ClientConnectorError as e:
            raise ConnectionError(e)

        if r.status != 200:
            raise BadStatusError(r.status)

        try:
            response = yield from r.json(loads=json.loads)
        except ValueError:
            raise BadJsonError('Invalid received json from node.')

        if not id_:
            self._id += 1
        try:
            return response['result']
        except KeyError:
            raise BadResponseError(response['error']['message'],
                                   response['error']['code'])


class AsyncIOIPCClient(BaseAsyncIOClient, RpcMixin):
    """Creates AsyncIOIPCClient client to communicate via IPC.

    :param reader: Instance of the stream reader
    :type reader: :class:`asyncio.streams.StreamReader`

    :param writer: Instance of the stream writer
    :type writer: :class:`asyncio.streams.StreamWriter`

    :param unix_path: Unix domain path
    :type unix_path: str

    :param timeout: An optional total time of timeout call
    :type timeout: int

    :param loop: An optional *event loop* instance
                 (uses :func:`asyncio.get_event_loop` if not specified).
    :type loop: :ref:`EventLoop<asyncio-event-loop>`

    :return: :class:`AsyncIOIPCClient` instance.
    """

    def __init__(self, reader, writer, unix_path, timeout=60, *, loop=None):
        self._reader = reader
        self._writer = writer
        self._timeout = timeout
        self._unix_path = unix_path
        self._id = 1
        self._loop = loop or asyncio.get_event_loop()
        self._lock = asyncio.Semaphore(1, loop=self._loop)

    @asyncio.coroutine
    def rpc_call(self, method, params=None, id_=None):
        params = params or []
        data = {
            'jsonrpc': '2.0',
            'method': method,
            'params': params,
            'id': id_ or self._id,
        }
        with (yield from self._lock):
            self._writer.write(json.dumps(data).encode('utf-8'))
            b = yield from self._reader.readline()
            if not b:
                tried = 0
                while True:
                    try:
                        new_client = yield from create_ethereum_client(
                            self._unix_path, self._timeout)
                    except Exception:
                        tried += 1
                        if tried == _reconnect_times:
                            raise ConnectionError('Didn\'t receive any data, '
                                                  'connection refused.')
                    else:
                        break
                self._id = new_client._id
                self._reader = new_client._reader
                self._writer = new_client._writer
                self._lock = new_client._lock
                return (yield from self.rpc_call(method, params, id_))

        try:
            response = json.loads(b.decode('utf-8'))
        except ValueError:
            raise BadJsonError('Invalid received json from node.')

        if not id_:
            self._id += 1
        try:
            return response['result']
        except KeyError:
            raise BadResponseError(response['error']['message'],
                                   response['error']['code'])


@asyncio.coroutine
def create_ethereum_client(uri, timeout=60, *, loop=None):
    """Create client to ethereum node based on schema.

    :param uri: Host on ethereum node
    :type uri: str

    :param timeout: An optional total time of timeout call
    :type timeout: int

    :param loop: An optional *event loop* instance
                 (uses :func:`asyncio.get_event_loop` if not specified).
    :type loop: :ref:`EventLoop<asyncio-event-loop>`

    :return: :class:`BaseAsyncIOClient` instance.
    """
    if loop is None:
        loop = asyncio.get_event_loop()

    presult = urlparse(uri)
    if presult.scheme in ('ipc', 'unix'):
        reader, writer = yield from asyncio.open_unix_connection(presult.path,
                                                                 loop=loop)
        return AsyncIOIPCClient(reader, writer, uri, timeout, loop=loop)
    elif presult.scheme in ('http', 'https'):
        tls = presult.scheme[-1] == 's'
        netloc = presult.netloc.split(':')
        host = netloc.pop(0)
        port = netloc.pop(0) if netloc else (443 if tls else 80)
        return AsyncIOHTTPClient(host, port, tls, timeout, loop=loop)
    else:
        raise RuntimeError('This scheme does not supported.')
